components/camel-rocketmq/src/main/java/org/apache/camel/component/rocketmq/RocketMQProducer.java

```diff
@@ -0,0 +1,231 @@
+    public RocketMQProducer(RocketMQEndpoint endpoint) {
+        super(endpoint);
+    }
+
+    @Override
+    public RocketMQEndpoint getEndpoint() {
+        return (RocketMQEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        if (!isRunAllowed()) {
+            if (exchange.getException() == null) {
+                exchange.setException(new RejectedExecutionException());
+            }
+            callback.done(true);
+            return true;
+        }
+        try {
+            LOG.trace("Exchange Pattern {}", exchange.getPattern());
+            if (exchange.getPattern().isOutCapable()) {
+                return processInOut(exchange, callback);
+            } else {
+                return processInOnly(exchange, callback);
+            }
+        } catch (Throwable e) {
+            exchange.setException(e);
+            callback.done(true);
+            return true;
+        }
+    }
+
+    protected boolean processInOut(final Exchange exchange, final AsyncCallback callback)
+            throws RemotingException, MQClientException, InterruptedException, NoTypeConversionAvailableException {
+        org.apache.camel.Message in = exchange.getIn();
+        Message message = new Message();
+        message.setTopic(in.getHeader(RocketMQConstants.OVERRIDE_TOPIC_NAME, () -> getEndpoint().getTopicName(), String.class));
+        message.setTags(in.getHeader(RocketMQConstants.OVERRIDE_TAG, () -> getEndpoint().getSendTag(), String.class));
+        message.setBody(exchange.getContext().getTypeConverter().mandatoryConvertTo(byte[].class, exchange, in.getBody()));
+        message.setKeys(in.getHeader(RocketMQConstants.OVERRIDE_MESSAGE_KEY, "", String.class));
+        initReplyManager();
+        String generateKey = GENERATE_MESSAGE_KEY_PREFIX + getEndpoint().getCamelContext().getUuidGenerator().generateUuid();
+        message.setKeys(Arrays.asList(Optional.ofNullable(message.getKeys()).orElse(""), generateKey));
+        LOG.debug("RocketMQ Producer sending {}", message);
+        mqProducer.send(message, new SendCallback() {
+
+            @Override
+            public void onSuccess(SendResult sendResult) {
+                if (!SendStatus.SEND_OK.equals(sendResult.getSendStatus())) {
+                    exchange.setException(new SendFailedException(sendResult.toString()));
+                    callback.done(false);
+                }
+                if (replyManager == null) {
+                    LOG.warn("replyToTopic not set! Will not wait for reply.");
+                    callback.done(false);
+                    return;
+                }
+                replyManager.registerReply(replyManager, exchange, callback, generateKey,
+                        getEndpoint().getRequestTimeoutMillis());
+            }
+
+            @Override
+            public void onException(Throwable e) {
+                replyManager.cancelMessageKey(generateKey);
```

Comment:
maybe use try .. finally to ensure callback is called if replymanager for some strange reason throw exception as callback MUST be called